Kudu.Core/Jobs/BaseJobRunner.cs

```diff
@@ -72,17 +71,65 @@ protected string WorkingDirectory
 
         protected JobSettings JobSettings { get; set; }
 
-        private static int CalculateHashForJob(string jobBinariesPath)
+        internal static bool JobDirectoryHasChanged(
+            Dictionary<string, FileInfoBase> sourceDirectoryFileMap, 
+            Dictionary<string, FileInfoBase> workingDirectoryFileMap, 
+            Dictionary<string, FileInfoBase> cachedSourceDirectoryFileMap)
         {
-            var updateDatesString = new StringBuilder();
-            DirectoryInfoBase jobBinariesDirectory = FileSystemHelpers.DirectoryInfoFromDirectoryName(jobBinariesPath);
+            // enumerate all source directory files, and compare against the files
+            // in the working directory (i.e. the cached directory)
+            FileInfoBase foundEntry = null;
+            foreach (var entry in sourceDirectoryFileMap)
+            {  
+                if (workingDirectoryFileMap.TryGetValue(entry.Key, out foundEntry))
+                {
+                    if (entry.Value.LastWriteTimeUtc != foundEntry.LastWriteTimeUtc)
+                    {
+                        // source file has changed since we last cached it, or a source
+                        // file has been modified in the working directory.
+                        return true;
+                    }
+                }
+                else
+                {
+                    // A file exists in source that doesn't exist in our working
+                    // directory. This is either because a file was actually added
+                    // to source, or a file that previously existed in source has been
+                    // deleted from the working directory.
+                    return true;
+                }
+            }
+
+            // if we've previously run this check in the current process,
+            // look for any file deletions by ensuring all our previously
+            // cached entries are still present
+            if (cachedSourceDirectoryFileMap != null)
+            {
+                foreach (var entry in cachedSourceDirectoryFileMap)
+                {
+                    if (!sourceDirectoryFileMap.TryGetValue(entry.Key, out foundEntry))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        internal static Dictionary<string, FileInfoBase> GetJobDirectoryFileMap(string sourceDirectory)
+        {
+            Dictionary<string, FileInfoBase> fileMap = new Dictionary<string, FileInfoBase>();
```

Comment:
Needs to be case insensitive. Someone could update a file and change its case, but it's still the same file.